### PR TITLE
Parse arbitrary precision integers ..

### PR DESCRIPTION
**Pull request by hhugo**
.. and allow any char of [A-Za-z] as modifier for ints (previously 'l','L','n') and floats. 
This give more freedom to ppx rewritters (what about a ppx for zarith or literal for complex)

Checks are performed when translating from Parsetree to Typedtree.
Unknow_literal is raised if the modifier is not recognized ([lLn]?)
Integer_overflow is raised as before.

### CODE REVIEW

File: parsing/parser.mly [hunk 1]
```diff
@@ -2096,7 +2079,7 @@ class_longident:
 toplevel_directive:
     SHARP ident                 { Ptop_dir($2, Pdir_none) }
   | SHARP ident STRING          { Ptop_dir($2, Pdir_string (fst $3)) }
-  | SHARP ident INT             { Ptop_dir($2, Pdir_int $3) }
+  | SHARP ident INT             { Ptop_dir($2, Pdir_int (int_of_string $3)) }
```
Comment (hhugo) [inline parsing/parser.mly]: This may fail

File: parsing/lexer.mll [hunk 2]
```diff
@@ -330,31 +318,11 @@ rule token = parse
...
   | uppercase_latin1 identchar_latin1 *
       { warn_latin1 lexbuf; UIDENT(Lexing.lexeme lexbuf) }
-  | int_literal
-      { try
-          INT (cvt_int_literal (Lexing.lexeme lexbuf))
-        with Failure _ ->
-          raise (Error(Literal_overflow "int", Location.curr lexbuf))
-      }
+  | int_literal { INT (Lexing.lexeme lexbuf) }
+  | (int_literal as num) (identchar + as modifier)
```
Comment (hhugo) [inline parsing/lexer.mll]: a 'modifier' could simply be ['a'-'z''A'-Z']

**Comment by Octachron** [comment]
This patch looks really interesting. Would it  not make sense to have the same kind of support for float litteral? I can see some applications for qualified float litterals: complex number, single or quadruple precision litteral for instance. Most of these applications are not general enough to be supported within the compiler but would make nice ppx extensions.

File: parsing/printast.ml [hunk 3]
```diff
@@ -887,7 +886,8 @@ and directive_argument i ppf x =
   match x with
   | Pdir_none -> line i ppf "Pdir_none\n"
   | Pdir_string (s) -> line i ppf "Pdir_string \"%s\"\n" s;
-  | Pdir_int (i) -> line i ppf "Pdir_int %d\n" i;
```
Comment (hhugo) [inline parsing/printast.ml]: There was an error here because the argument `i` of the function was shadowed by the pattern matching,

**Comment by hhugo** [comment]
@Octachron, I've updated the PR to add modifier on floats.

File: parsing/parsetree.mli [hunk 4]
```diff
@@ -14,6 +14,12 @@
 
 open Asttypes
 
+type constant =
+    PConst_int of string * char option
```
Comment (bobot) [inline parsing/parsetree.mli:18]: The minus sign is kept inside the string. So there is a lot of string processing for this first character. Is it not easier to remove at lexing the minus sign from the string and keep the information in a separate bool?

**Comment by samoht** [comment]
It could maybe be helpful if you add a few tests exercising the new code?

File: parsing/printast.ml (hunk 3 above)
Comment (hhugo) [inline parsing/printast.ml]: 15 years old typo !!

**Comment by Drup** [comment]
I like the patch. I read the implementation and I think it's good.

I agree with @bobot on the minus sign.

I think there should be a function from Parsetree.constant to Asttypes.constant that do not need the typing env. It would be very useful for ppx writer. Returning an option would work well.

File: parsing/lexer.mll [hunk 5]
```diff
@@ -330,29 +315,21 @@ rule token = parse
...
   | uppercase_latin1 identchar_latin1 *
       { warn_latin1 lexbuf; UIDENT(Lexing.lexeme lexbuf) }
-  | int_literal
-      { try
-          INT (cvt_int_literal (Lexing.lexeme lexbuf))
-        with Failure _ ->
-          raise (Error(Literal_overflow "int", Location.curr lexbuf))
+  | int_literal { INT (Lexing.lexeme lexbuf, None) }
+  | int_literal literal_modifier
+      {
```
Comment (xavierleroy) [inline parsing/lexer.mll]: This would really benefit from using ocamllex's named sub-strings:

| (int_literal as lit) (literal_modifier as modif) { INT (lit, Some modif) }

**Comment by hhugo** [comment]
I would prefer to keep the whole literal (minus sign included) inside the string. In most cases, it does not add any string processing as we already parse every int literal as negative numbers. (prepending  `"-"` if not  already negative, https://github.com/ocaml/ocaml/pull/170/files#diff-4408a3671b05f01d8b0dbe9df336e08bL173)

**Comment by hhugo** [comment]
Do not merge until camlp4 PR is ready https://github.com/ocaml/camlp4/pull/90

**Comment by hhugo** [comment]
PR ready for camlp4 https://github.com/ocaml/camlp4/pull/91

File: utils/misc.ml [hunk 6]
```diff
@@ -180,6 +180,17 @@ let no_overflow_mul a b = b <> 0 && (a * b) / b = a
 
 let no_overflow_lsl a k = 0 <= k && k < Sys.word_size && min_int asr k <= a && a <= max_int asr k
 
+module Int_literal_converter = struct
+  let cvt_int_aux str neg of_string =
+    if String.length str = 0 || str.[0]= '-'
+    then of_string str
+    else neg (of_string ("-" ^ str))
```
Comment (bobot) [inline utils/misc.ml:196]: Could you add a comment for why you need this complicated function and a link to the related bts bug?

File: typing/typecore.ml [hunk 7]
```diff
@@ -261,6 +263,40 @@ let type_constant = function
   | Const_int64 _ -> instance_def Predef.type_int64
   | Const_nativeint _ -> instance_def Predef.type_nativeint
 
+(* To convert integer literals, allowing max_int + 1 (PR#4210) *)
```
Comment (bobot) [inline typing/typecore.ml]: I never understood this comment (which is in the original code) since it allows to _fail_ well on `max_int+1`. What do you think of it?

**Comment by gasche** [comment]
This one looks ready to merge. What is the status of ocaml/camlp4#90 ?

**Comment by hhugo** [comment]
Rebased on top of trunk.
Ready to merge unless one wants to add tests as proposed by @samoht

**Comment by gasche** [comment]
Tests are always good. I'd love tests!

File: parsing/printast.ml [hunk 8]
```diff
@@ -51,15 +51,14 @@ let fmt_string_loc f x =
 
 let fmt_constant f x =
   match x with
-  | Const_int (i) -> fprintf f "Const_int %d" i;
-  | Const_char (c) -> fprintf f "Const_char %02x" (Char.code c);
-  | Const_string (s, None) -> fprintf f "Const_string(%S,None)" s;
-  | Const_string (s, Some delim) ->
+  | PConst_int (i,None) -> fprintf f "Const_int %s" i;
```
Comment (gasche) [inline parsing/printast.ml]: Given that the format here is apparently meant to give re-parseability, you should reprint `PConst` rather than `Const`.

**Comment by gasche** [comment]
If you write tests, it would be nice to include one checking that `-1073741824` is still correctly parsed as an integer literal (This corresponds to min_int for 32 machines.), to avoid [the issue described by this comment in MPR#4210](http://caml.inria.fr/mantis/view.php?id=4210#c3921).

**Comment by gasche** [comment]
I have one worry with this patch: there is an ambiguity (resolved by the longest-lexeme rule) between literal modifiers and hexadecimal literals. For example, `32f` would be parsed as a literal modifier, but `0x32f` is a valid hexadecimal literal.

Wouldn't it be safer (for now) to either restrict literal modifiers to the non-hexadecimal cases (but this wouldn't be consistent with `nlL` which are valid for hexadecimal literals), or to letters strictly after `f` or `F`?

**Comment by hhugo** [comment]
How safer would that be ?
It seams to be similar to `1.2e` and `1.2e3`

**Comment by dbuenzli** [comment]
> .. and allow any char of [A-Za-z] as modifier for ints

This may break existing programs, right ?

**Comment by gasche** [comment]
Not directly, in the sense that programs that relied on `123foo` to parse were made illegal by #155 . That said, #155 has not yet been part of a release, so both changes together will break some programs, yes.

**Comment by bobzhang** [comment]
So this patch is divided into two parts:
1. make parsetree using string literals (more loyal to user input)
2. extending lexer to support arbitrary modifiers
   
   With regard to 1, can we make changes to Asttypes.constant directly instead of creating a new type, Asttypes.constant is also used by Lambda, I am currently working on a project compiling lambda to readable javascript in the module level (still in the early days http://bobzhang.github.io/js-demo/), we could benefit such changes as well.

**Comment by hhugo** [comment]
I've just rebased this PR onto trunk. Are people happy the current state ?

@bobzhang, you proposal of updating `Asttypes.constant` and not introducing a new type does not really make sense to me. One might still want to propagate constant as string to lambda but I would not do it inside this PR.

**Comment by bobzhang** [comment]
my 2 cents: introducing another type actually increases the complexity of the patch (you have to update typeast and untypeast), I would say updating `Asttypes.constant` is good for both

**Comment by hhugo** [comment]
The two types have very distinct (/ opposite )  purposes.
- `Parsetree.constant` have very little constraint to give more freedom to ppx rewriters
- `Asttypes.constant` enforce that constants are well-formed (integer should be int, int32, int64 .., and should not overflow).

Unifying the two types would make the patch more complex but feel free to prove me wrong

**Comment by Drup** [comment]
As a ppx writer, I really like the distinction between the two types and I feel like it's useful.

**Comment by damiendoligez** [comment]
@hhugo you need to add a line to `Changes`.

**Comment by hhugo** [comment]
Change log updated.

note that https://github.com/hhugo/ocaml/commit/7681a73f5e66214fa0d701c91654bf5cfd8caa41 should _not_ be merged with the rest

File: typing/typecore.ml [hunk 9]
```diff
@@ -4202,12 +4239,18 @@ let report_error env ppf = function
       fprintf ppf
         "@[This form is not allowed as the type of the inlined record could \
          escape.@]"
-  | Unrefuted_pattern pat ->
+ | Unrefuted_pattern pat ->
```
Comment (damiendoligez) [inline typing/typecore.ml]: Why did you change the indentation here? This looks wrong.

File: Changes [hunk 10]
```diff
@@ -298,6 +298,9 @@ Features wishes:
   (Florian Angeletti)
 - GPR#252: improve build instructions in MSVC Windows README
   (Philip Daian)
+- GRP#170: Parse arbitrary precision integers and accept a single [A-Za-z]
```
Comment (damiendoligez) [inline Changes]: typo on GPR

**Comment by damiendoligez** [comment]
Good to merge once the typos are fixed.

**Comment by hhugo** [comment]
@damiendoligez, should I rebase and remove the temporary commit related to `opam pin` ?

**Comment by gasche** [comment]
Yes, rebasing is always a good idea to keep the commit history clean (we _do_ use bisect etc.), see [Contributing: Clean patch series](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#clean-patch-series).

**Comment by hhugo** [comment]
PR rebased and cleared.
Note that the travis will fail because camlp4 is not pinned anymore.

File: Changes [hunk 11]
```diff
@@ -328,6 +328,9 @@ Features wishes:
   (Florian Angeletti)
 - GPR#252: improve build instructions in MSVC Windows README
   (Philip Daian)
+* GPR#170: Parse arbitrary precision integers and accept a single [A-Za-z]
```
Comment (gasche) [inline Changes]: Could you maybe add a line to clarify why this can break existing programs? It would probably suffice to just show an example: `let a = 1 in (+) 123a` was accepted but is now rejected. This would reassure users that, unless they did horrendous stuff, they should be quite safe.

(You should feel free to squash your Changelog commits together.)

**Comment by gasche** [comment]
This looks good to merge, but I'm still mildly non-plussed by the interaction between modifiers in the a-f range and hexa literals. I assigned @damiendoligez so that he takes care of merging.

**Comment by hhugo** [comment]
@gasche, thinking more about it, I now agree with your concern.
How do you feel about restricting the range for modifiers to ['g'-'z''G'-'Z'] ?

**Comment by gasche** [comment]
I think that this would be a good, safe thing to do -- we can always extend the feature later with more experience.

**Comment by damiendoligez** [comment]
Merged with [g-zG-Z].

**Comment by alainfrisch** [comment]
The new test min_int.ml cannot be compiled on 32-bit architectures.  I've disabled it for now.

If someone insists on keeping it, please tweak the test suite to exclude this test (or create the 32-bit counterpart).

**Comment by gasche** [comment]
We should have a warning on integer literals that do not fit on 32 machines, in the same spirit as the marshalling flag.

**Comment by alainfrisch** [comment]
Note that there is already "ocamlc -compat-32", which triggers an error when the bytecode is not compatible with 32-bit.

**Comment by damiendoligez** [comment]
I've "fixed" the test. Instead of adding heavy shell plumbing to compile a different source depending on the platform, I just changed it to the 32-bit `min_int`. This means the test is mostly useless on 64-bit, but the 32-bit platforms will test the corner case.

**Comment by damiendoligez** [comment]
@hhugo [camlp4](https://github.com/ocaml/camlp4) is broken by this change. Can you have a look?

**Comment by bobzhang** [comment]
curious do we have any ppx make use of such extension

**Comment by Drup** [comment]
@bobzhang https://github.com/ocaml/Zarith/pull/4
